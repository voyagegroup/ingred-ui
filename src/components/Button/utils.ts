import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
