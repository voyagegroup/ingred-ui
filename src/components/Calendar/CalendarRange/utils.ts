import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
