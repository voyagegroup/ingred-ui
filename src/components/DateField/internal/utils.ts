export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
