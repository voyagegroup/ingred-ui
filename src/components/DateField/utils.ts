import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
