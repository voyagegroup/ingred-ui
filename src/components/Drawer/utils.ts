/**
 * 相対値をpx値に変換するヘルパー関数
 * @param value - 変換する値（数値またはvw/vh/%/px等の単位付き文字列）
 * @param direction - Drawerの方向（変換時の基準値決定に使用）
 * @returns px値（数値）
 */
export const convertToPixels = (value: string | number, direction: "left" | "right" | "bottom"): number => {
  if (typeof value === "number") {
    return value; // 数値はpx扱い
  }
  
  // 文字列の場合は単位チェック
  const numericValue = parseFloat(value);
  
  if (value.includes("vw")) {
    return (numericValue / 100) * window.innerWidth;
  }
  if (value.includes("vh")) {
    return (numericValue / 100) * window.innerHeight;
  }
  if (value.includes("%")) {
    // %の場合は親要素基準だが、ここではviewport基準として扱う
    const isHorizontal = direction === "left" || direction === "right";
    return (numericValue / 100) * (isHorizontal ? window.innerWidth : window.innerHeight);
  }
  if (value.includes("px")) {
    return numericValue;
  }
  
  // 単位なしの文字列は警告を出す
  if (!isNaN(numericValue)) {
    console.warn(`Drawer: 文字列サイズには単位を指定してください。"${value}" → "${value}px" を推奨`);
    return numericValue; // 一応数値として扱う
  }
  
  // 無効な値の場合はエラー警告とフォールバック
  console.error(`Drawer: 無効なサイズ値です: "${value}". デフォルト値を使用します。`);
  return 400; // フォールバック
}; 