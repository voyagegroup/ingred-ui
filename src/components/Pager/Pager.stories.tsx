import * as React from "react";
import styled from "styled-components";
import Pager, { useFilterState, getFilteredItems, PagerProps } from "./";
import { Story } from "@storybook/react/types-6-0";
import Typography from "../Typography";
import Spacer from "../Spacer";

const Container = styled.div`
  padding: ${({ theme }) => theme.spacing * 3}px;
  background-color: ${({ theme }) => theme.palette.background.dark};
`;

const RowContainer = styled.div<{ minHeight?: string; flex?: boolean }>`
  display: ${({ flex }) => (flex ? "flex" : "block")};
  align-items: ${({ flex }) => (flex ? "flex-start" : "normal")};
  margin: ${({ theme }) => theme.spacing * 3}px;
  padding: ${({ theme }) => theme.spacing * 3}px;
  background-color: ${({ theme }) => theme.palette.background.default};
  min-height: ${({ minHeight }) => minHeight || "0"};
  flex-wrap: wrap;
`;

export default {
  title: "Components/Data Display/Pager",
  component: Pager,
};

export const Example: Story<PagerProps> = () => {
  const mockArray: number[] = [...Array(61)].map((_, i) => i);

  // MEMO: `usefilterState()` & `getFilteredItems()` is available from ingred-ui.
  // e.g. filterState = { per: 10, index: 3 }
  const [filterState, setFilterState] = useFilterState(10, 1); // (per, index)
  const filteredItems = getFilteredItems(mockArray, filterState);

  const handleChangePager = (index: number) => {
    setFilterState({ ...filterState, index });
  };

  return (
    <>
      <ul>
        {filteredItems.map((num) => (
          <li key={num}>{num}</li>
        ))}
      </ul>
      <Pager
        per={filterState.per}
        total={mockArray.length}
        index={filterState.index}
        onClick={handleChangePager}
      />
    </>
  );
};

export const DesignSamples: Story = () => (
  <Container>
    <RowContainer flex>
      <div>
        <Typography weight="bold" size="xxl">
          Can&apos;t Forward
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={2} index={2} onClick={() => {}} />
      </div>
      <Spacer pr={5} />
      <div>
        <Typography weight="bold" size="xxl">
          Can&apos;t Back
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={2} index={1} onClick={() => {}} />
      </div>
      <Spacer pr={5} />
      <div>
        <Typography weight="bold" size="xxl">
          Can Both
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={3} index={2} onClick={() => {}} />
      </div>
    </RowContainer>

    <RowContainer flex>
      <div>
        <Typography weight="bold" size="xxl">
          Left Three Dots
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={7} index={7} onClick={() => {}} />
      </div>
      <Spacer pr={5} />
      <div>
        <Typography weight="bold" size="xxl">
          Both Three Dots
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={7} index={4} onClick={() => {}} />
      </div>
      <Spacer pr={5} />
      <div>
        <Typography weight="bold" size="xxl">
          Rignt Three Dots
        </Typography>
        <Spacer mb={3} />
        <Pager per={1} total={7} index={1} onClick={() => {}} />
      </div>
    </RowContainer>
  </Container>
);
