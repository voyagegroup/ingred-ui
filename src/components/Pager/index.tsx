export { default, useFilterState, getFilteredItems } from "./Pager";
export type { FilterState, PagerProps } from "./Pager";
