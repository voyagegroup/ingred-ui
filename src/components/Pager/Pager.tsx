import * as React from "react";

import { NumberButton } from "./internal/NumberButton";
import { ArrowButton } from "./internal/ArrowButton";
import { PagerButtonsWithDottedLine } from "./internal/PagerButtonsWithDottedLine";
import Flex from "../Flex";

export type PagerProps = {
  per: number;
  total: number;
  index?: number;
  onClick: (index: number) => void;
};

const Pager = React.forwardRef<HTMLDivElement, PagerProps>(function Pager(
  { per, total, index = 1, onClick },
  ref,
) {
  const pagerCount = Math.ceil(total / per);
  const isFirst = index === 1 || total === 0;
  const isLast = index === pagerCount || total === 0;

  const handleClick = (index: number) => () => onClick(index);

  return (
    <Flex ref={ref} display="flex" alignItems="center">
      <ArrowButton
        isRight={false}
        disabled={isFirst}
        onClick={handleClick(index - 1)}
      />
      {pagerCount < 7 ? (
        [...Array(pagerCount)].map((_, i) => (
          <NumberButton
            key={i} // eslint-disable-line react/no-array-index-key
            index={i + 1}
            isActiveIndex={index === i + 1}
            onClick={handleClick(i + 1)}
          />
        ))
      ) : (
        <PagerButtonsWithDottedLine
          pagerCount={pagerCount}
          index={index}
          onClick={onClick}
        />
      )}
      <ArrowButton isRight disabled={isLast} onClick={handleClick(index + 1)} />
    </Flex>
  );
});

export type FilterState = {
  per: number;
  index: number;
};

export function getFilteredItems<T>(items: T[], { per, index }: FilterState) {
  return items.slice((index - 1) * per, index * per);
}

export const useFilterState = (per = 100, index = 1) =>
  React.useState<FilterState>({ per, index });

export default Pager;
