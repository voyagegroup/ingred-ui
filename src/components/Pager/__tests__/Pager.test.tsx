import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import Pager from "../Pager";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Pager component testing", () => {
  afterEach(cleanup);

  test("Pager", () => {
    const { asFragment } = renderWithThemeProvider(
      <Pager per={10} total={100} index={1} onClick={jest.fn} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
