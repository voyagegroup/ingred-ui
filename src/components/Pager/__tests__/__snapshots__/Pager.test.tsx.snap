// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager component testing Pager 1`] = `
<DocumentFragment>
  <div
    class="sc-jRQBWg dEAIxE"
    display="flex"
  >
    <button
      class="sc-dkPtRN ikClgn"
      disabled=""
      type="button"
    >
      <div
        class="sc-hKwDye cPkzPO"
        size="18"
      >
        <svg
          transform="rotate(90)"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0,0H24V24H0Z"
            fill="none"
          />
          <path
            d="M13.414,16.886,7.757,11.228,9.644,9.343l3.771,3.772,3.771-3.772,1.887,1.885Z"
            fill="#B3BAC1"
            transform="translate(-1.414 -1.114)"
          />
        </svg>
      </div>
    </button>
    <button
      class="sc-bdvvtL bWclDf"
      disabled=""
      type="button"
    >
      <span
        class="sc-gsDKAQ eaLEXZ"
        color="#FFFFFF"
        font-size="12px"
      >
        1
      </span>
    </button>
    <button
      class="sc-bdvvtL dUUjWf"
      type="button"
    >
      <span
        class="sc-gsDKAQ iDsTVX"
        color="#0B82F4"
        font-size="12px"
      >
        2
      </span>
    </button>
    <button
      class="sc-bdvvtL dUUjWf"
      type="button"
    >
      <span
        class="sc-gsDKAQ iDsTVX"
        color="#0B82F4"
        font-size="12px"
      >
        3
      </span>
    </button>
    <button
      class="sc-bdvvtL dUUjWf"
      type="button"
    >
      <span
        class="sc-gsDKAQ iDsTVX"
        color="#0B82F4"
        font-size="12px"
      >
        4
      </span>
    </button>
    <button
      class="sc-bdvvtL dUUjWf"
      type="button"
    >
      <span
        class="sc-gsDKAQ iDsTVX"
        color="#0B82F4"
        font-size="12px"
      >
        5
      </span>
    </button>
    <div
      class="sc-eCImPb bUSdLm"
    >
      ...
    </div>
    <button
      class="sc-bdvvtL dUUjWf"
      type="button"
    >
      <span
        class="sc-gsDKAQ iDsTVX"
        color="#0B82F4"
        font-size="12px"
      >
        10
      </span>
    </button>
    <button
      class="sc-dkPtRN dRYOeP"
      type="button"
    >
      <div
        class="sc-hKwDye cPkzPO"
        size="18"
      >
        <svg
          transform="rotate(-90)"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0,0H24V24H0Z"
            fill="none"
          />
          <path
            d="M13.414,16.886,7.757,11.228,9.644,9.343l3.771,3.772,3.771-3.772,1.887,1.885Z"
            fill="#0B82F4"
            transform="translate(-1.414 -1.114)"
          />
        </svg>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
